app/vmselect/promql/aggr.go

```diff
@@ -129,21 +139,7 @@ func aggrFuncExt(afe func(tss []*timeseries, modifier *metricsql.ModifierExpr) [
 		m[k] = tss
 	}
 	bbPool.Put(bb)
-
-	srcTssCount := 0
-	dstTssCount := 0
-	rvs := make([]*timeseries, 0, len(m))
-	for _, tss := range m {
-		rv := afe(tss, modifier)
-		rvs = append(rvs, rv...)
-		srcTssCount += len(tss)
-		dstTssCount += len(rv)
-		if dstTssCount > 2000 && dstTssCount > 16*srcTssCount {
```

Comment:
This code already contains the protection from the explosion on the number of time series generated by `count_values()`, e.g. it prevents from generating more than `16x(source time series)` for the output time series if the number of output time series exceeds 2000. It looks like this protection didn't work as expected in the observed case when `vmselect` memory usage was steadily increasing to the maximum available memory during 30 minutes when the `count_values()` query was executed. Obviously, the number of output time series was bigger than 2000 in this case. But in this case the `dstTssCount > 16*srcTssCount` should trigger early if `count_values()` was generating more than 16 new time series per each input series. On the other hand, if the number of source time series was big (e.g. millions), then the memory usage should grow much faster, e.g. `vmselect` should reach OOM crash in seconds or a few minutes at max, not 30 minutes.

Reply:
Do you suggest I keep this code for all aggregations, not for `count_values` only?

Reply:
No, it is better to use this check specifically for `count_values()` function, since other aggregate functions cannot produce more output time series than the input time series.